### DISCUSSION TITLE
Using capabilities of Fipy to employ (define and solve) locally averaged Navier-Stokes equations in a coupled CFD-DEM simulation

### DISCUSSION THREAD

**Discussion by hadimorovvatju** [usnistgov/fipy#966] [Q&A] [answered]
Hi, 

I hope this request post finds you well. In order to implement unresolved three-dimensional CFD-DEM (using "Fipy" python Lib-Itasca PFC commercial software) coupling, I need to use capabilities of "Fipy" to employ (define and solve) locally averaged Navier-Stokes equations. Unfortunately, my CFD knowledge does not suffice. So at this point, I seriously need help and any helps would be greatly appreciated. All I have is an example in Itasca PFC 7.0 documentation of implementing Darcy's law using "Fipy" to model low Reynolds number porous flow in a granular material.[This is the link of Darcy flow example](https://docs.itascacg.com/pfc700/pfc/docproject/source/manual/optional_features/cfd_option/cfd_darcy.html?highlight=darcy%20flow)
Now, I need to implement locally averaged Navier-Stokes equations which are attached to the end of the post. As for setting boundary conditions, one may impose constant pressures at both inlet (non-zero) and outlet (0) of the sample in order to apply hydraulic gradients and free-slip condition on the rest. How may I formulate and solve these equations using Fipy by manipulating the example code mentioned earlier?

Thank you in advance for your consideration. Here are Navier-Stokes equations mathematical formulation from 2 different articles:
![DOI: 10.1016/S1001-6058(14)60046-9](https://github.com/usnistgov/fipy/assets/100556858/212822a5-73af-42fa-85e3-eae90c30d3e9)
![https://doi.org/10.25560/60850](https://github.com/usnistgov/fipy/assets/100556858/e3870790-1d9f-4570-aa20-1e8ed059a4c3)

**Comment by wd15** [comment, marked as answer]
Hi @hadimorovvatju, here are a few previous answers to similar questions that might help you get started.

 - [Stackoverflow question for full cavity flow](https://stackoverflow.com/questions/66243908/fipy-2d-navier-stokes-implementation-problem-with-derivatives-in-one-direction)
 - Maybe just start with Stokes and build from there, see [this](https://stackoverflow.com/questions/60712013/boundary-conditions-for-stokes-flow-around-a-sphere-using-fipy).
 - [An implementation](https://github.com/guyer/phasefieldbenchmark-5/blob/master/problem5.py) for this [specified problem](https://pages.nist.gov/pfhub/benchmarks/benchmark5-hackathon.ipynb/).

Here's how I would approach the problem,

 - Ensure the [original code](https://docs.itascacg.com/pfc700/pfc/docproject/source/manual/optional_features/cfd_option/cfd_darcy.html?highlight=darcy%20flow) runs and makes sense. It's the test bed for the following steps. 
 - Next I would take one of the two Stokes examples above and slowly adapt them (making sure they work at each step) to be close to the problem that I'm interested in and see which one works best.  For example, only add in the gravity term only once the initial implementation converges and makes sense.
 - Try and make the original code and the Stokes example get close in their limits of applicability.
 - Assuming that Stokes works then add in Transient terms. It might not hurt to add in Transient terms with the above steps. They tend to make for a more stable solution.
 - Assuming that the above all works then add in the convection terms.
 - Do not attempt anything in 3D until 2D is implemented and understood.

**Reply by hadimorovvatju** [reply]
Thank you @wd15 for recommending relevant sources, plus sharing your invaluable vision, knowledge and directional methodology with us. I'm gonna try my hardest to employ your systematic approach in order to implement the solution.

Thank you again
